zebedee-cms/src/main/java/com/github/onsdigital/zebedee/keyring/io/CollectionKeyReadWriterImpl.java

```diff
@@ -0,0 +1,147 @@
+package com.github.onsdigital.zebedee.keyring.io;
+
+import com.github.onsdigital.zebedee.keyring.CollectionKey;
+import com.github.onsdigital.zebedee.keyring.KeyringException;
+import com.google.gson.Gson;
+import com.google.gson.GsonBuilder;
+import org.apache.commons.lang3.StringUtils;
+
+import javax.crypto.Cipher;
+import javax.crypto.CipherInputStream;
+import javax.crypto.CipherOutputStream;
+import javax.crypto.SecretKey;
+import javax.crypto.spec.IvParameterSpec;
+import java.io.FileInputStream;
+import java.io.FileOutputStream;
+import java.io.InputStreamReader;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.nio.file.Paths;
+
+/**
+ * {@link java.io.File} based implementation of {@link CollectionKeyReadWriter}. Reads and writes
+ * {@link CollectionKey} objects to/from encrypted files on disk.
+ * <p><b>We strongly advised against using/extending this
+ * code for anything other than maintaining legacy functionality in Zebedee CMS.</b></p>For all other purposes it
+ * should be considered deprecated.
+ */
+public class CollectionKeyReadWriterImpl implements CollectionKeyReadWriter {
+
+    private static final String ALGORITHM = "AES/CBC/PKCS5Padding";
+
+    private Path keyringDir;
+    private SecretKey masterKey;
+    private IvParameterSpec masterIv;
+    private Gson gson;
+
+    /**
+     * <p>
+     * Create a new readWriter instance.
+     * </p>
+     *
+     * <p>
+     * <b>Important:</b> It is strongly recommended that the {@link SecretKey} & {@link IvParameterSpec}
+     * are not modified once they have been used to store collection keys. Changing these inputs will result in this
+     * object no longer being able to decrypt collection key files written with the previous encryption key.
+     * <p>
+     * Reverting the key/init vector values will resolve this issue for keys written with the original encryption
+     * parameters however, the same issue will now be the case for any key files written using the modified encryption
+     * params.
+     * </p>
+     *
+     * <p>
+     * <b>
+     * If a encryption parameters change is necessary it is strongly recommended you migrate all existing keys to
+     * use the updated encryption values.</b>
+     * </p>
+     *
+     * @param keyringDir the {@link Path} the read the {@link CollectionKey} files from.
+     * @param masterKey  the {@link SecretKey} to use when decrypting the {@link CollectionKey} files.
+     * @param masterIv   the {@link IvParameterSpec} to use when initializing the encryption {@link Cipher}.
+     */
+    public CollectionKeyReadWriterImpl(final Path keyringDir, final SecretKey masterKey, final IvParameterSpec masterIv) {
+        this.keyringDir = keyringDir;
+        this.masterKey = masterKey;
+        this.masterIv = masterIv;
+        this.gson = new GsonBuilder()
+                .registerTypeAdapter(CollectionKey.class, new CollectionKeySerializer())
+                .create();
+    }
```

Comment:
**General question:** I am not sure if I need to use a `FileLock` when reading/writing the key files. Keys are individual files and are only ever created, read or deleted never modified I am not sure the lock is necessary. The intention is to store all keys in an in memory cache which will further prevent race conditions to the file.

With this in mind (for now at least) I am planning to leave out the File locking. If it becomes necessary is can be added retrospectively with a reasonably small impact.

Reply:
I have added the file lock in the latest version (pushing soon). It will acquire an exclusive lock on the file so any requests attempting to access it will block until the lock is released.

This will impact performance but it is a necessary hit. However, I'm pretty confident the amount of concurrent activity on each key file will be low so I am hopefully it should be an unnoticeable performance hit.